core/src/main/resources/volcano-viz/viz-template.html

```diff
@@ -0,0 +1,323 @@
+            <button id="next-button" style="width: 80px; height: 40px; display:inline-block" disabled> next</button>
+        </div>
+        <ul id="rule-match-list" style="width: 300px; height: 600px; overflow: auto">
+        </ul>
+    </div>
+    <div class="column2">
+        <svg id="svg-canvas" width="1200px" height="800px"></svg>
+    </div>
+</div>
+
+<script id="js">
+
+    var allNodes = data.allNodes;
+    var ruleMatchSequence = data.ruleMatchSequence;
+    var ruleMatchInfoMap = data.ruleMatchInfoMap;
+    var nodeAddedInRule = data.nodeAddedInRule;
+
+    /*
+     * Graph data and D3 JS render related variables
+     */
+
+    // Create the input graph
+    var g = new dagreD3.graphlib.Graph({
+            compound: true
+        })
+        .setGraph({
+            rankdir: 'LR'
+        })
+        .setDefaultEdgeLabel(function () {
+            return {};
+        });
+
+    // Create the renderer
+    var render = new dagreD3.render();
+
+    // Set up an SVG group so that we can translate the final graph.
+    var svg = d3.select("svg");
+    var svgGroup = svg.append("g");
+
+    // Set up zoom support
+    var svg = d3.select("svg")
+        .attr("width", "1200px")
+        .attr("height", "800px")
+        .call(d3.zoom().on("zoom", function () {
+            svgGroup.attr("transform", d3.event.transform)
+        }));
+
+    /*
+     * Global State
+     */
+
+    var currentRuleID = undefined;
+
+    /*
+     * Event Handler functions
+     */
+
+    var setCurrentRule = (ruleMatchID) => {
+        // un-highlight previous entry
+        var prevRuleID = currentRuleID;
+        if (prevRuleID !== undefined) {
+            var prevRuleElement = document.getElementById(prevRuleID);
+            prevRuleElement.style.backgroundColor = "#FFFFFF";
+        }
+
+        currentRuleID = ruleMatchID;
+        document.getElementById('current-rule').innerText = currentRuleID;
+
+        var currentRuleElement = document.getElementById(currentRuleID);
+        currentRuleElement.style.backgroundColor = "#D3D3D3";
+
+        var ruleIndex = ruleMatchSequence.indexOf(currentRuleID);
+
+        document.getElementById("prev-button").disabled = false;
+        document.getElementById("next-button").disabled = false;
+
+        if (ruleIndex === 0) {
+            document.getElementById("prev-button").disabled = true;
+        }
+        if (ruleIndex === ruleMatchSequence.length - 1) {
+            document.getElementById("next-button").disabled = true;
+        }
+
+        createGraph(ruleMatchID);
+    }
+
+    var createGraph = (ruleMatchID) => {
+        var ruleMatchInfo = ruleMatchInfoMap[ruleMatchID]
+        console.log(ruleMatchInfo);
+
+        // remove previous rendered view and clear graph model
+        d3.select("svg g").selectAll("*").remove();
+        g.nodes().slice().forEach(nodeID => g.removeNode(nodeID));
```

Comment:
I think it would be better to update the graph instead of resetting it. We can leave it like this for the moment and change it when we add a functionality that would benefit from keeping the existing nodes.

Reply:
Agree, but we need to be careful and compare the old nodes with the new nodes. This could be tricky especially when set merge happens: one set is removed and relnodes from this set all go into another set. So re-drawing is eaiser.

One more thing on performance. This tool runs slow when the plan is ultra-large. I investigated and the major performance bottleneck is actually doing the **layout** instead of drawing the nodes. I can't recall the numbers but over 90% of the time is spent on the layout (`dagre`). 

I did an extensive investigation on different layout libraries, but `dagre` remains the best-looking one, so I decided to keep it although it's slow. One sad thing is that none of these layout libraries can do it incrementally.

Reply:
Indeed, merging sets would be quite difficult to handle.

How big is an ulta-large plan? Maybe we can provide a combo-box to switch between different layout algorithms (e.g., d3-force).